File: C++/source/Framework/Messaging/LightHouse/LightHouseTransportBroadcastLightPulseWorker.cpp
/***************************************************************************
 *   Copyright (C) 20016 Vidyasagara Guntaka                               *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/Messaging/LightHouse/LightHouseTransportBroadcastLightPulseWorker.h"
#include "Framework/Messaging/LightHouse/LightHouseToolKit.h"
#include "Framework/Messaging/LightHouse/LightHouseTransportBroadcastLightPulseMessage.h"
#include "Framework/Messaging/LightHouse/LightHouseTransportObjectManager.h"
#include "Framework/Messaging/LightHouse/LightHouseTransportTypes.h"
#include "Framework/Messaging/LightHouse/LightPulse.h"
#include "Framework/Utils/MulticastSenderSocket.h"

namespace WaveNs
{

LightHouseTransportBroadcastLightPulseWorker::LightHouseTransportBroadcastLightPulseWorker (LightHouseTransportObjectManager *pLightHouseTransportObjectManager)
    : WaveWorker (pLightHouseTransportObjectManager)
{
    addOperationMap (LIGHT_HOUSE_TRANSPORT_BROADCAST_LIGHT_PULSE, reinterpret_cast<WaveMessageHandler> (&LightHouseTransportBroadcastLightPulseWorker::broadcastLightPulseMessageHandler));

    m_pMulticastSenderSocket = new MulticastSenderSocket (LightHouseToolKit::getLightHouseMulticastIpAddress (), LightHouseToolKit::getLightHouseMulticastPort ());

    waveAssert (NULL != m_pMulticastSenderSocket, __FILE__, __LINE__);
    waveAssert (m_pMulticastSenderSocket->isValid (), __FILE__, __LINE__);
}

LightHouseTransportBroadcastLightPulseWorker::~LightHouseTransportBroadcastLightPulseWorker ()
{
}

void LightHouseTransportBroadcastLightPulseWorker::broadcastLightPulseMessageHandler (LightHouseTransportBroadcastLightPulseMessage *pLightHouseTransportBroadcastLightPulseMessage)
{
    trace (TRACE_LEVEL_DEVEL, "LightHouseTransportBroadcastLightPulseWorker::broadcastLightPulseMessageHandler : Entering ...");

    waveAssert (NULL != pLightHouseTransportBroadcastLightPulseMessage, __FILE__, __LINE__);

    LightPulse *pLightPulse = pLightHouseTransportBroadcastLightPulseMessage->getPLightPulse ();

    waveAssert (NULL != pLightPulse, __FILE__, __LINE__);

    string serializedLightPulseString;

    pLightPulse->serialize2 (serializedLightPulseString);

    m_pMulticastSenderSocket->send (serializedLightPulseString);

    pLightHouseTransportBroadcastLightPulseMessage->setCompletionStatus (WAVE_MESSAGE_SUCCESS);
    reply (pLightHouseTransportBroadcastLightPulseMessage);
}

}

File: Java/source/com/CxWave/Wave/Framework/Messaging/LightHouse/LightHouseTransportBroadcastLightPulseWorker.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.Messaging.LightHouse;

import com.CxWave.Wave.Framework.ObjectModel.WaveWorker;
import com.CxWave.Wave.Framework.ObjectModel.WaveWorkerPriority;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.Cardinality;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.OwnerOM;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.WorkerPriority;
import com.CxWave.Wave.Framework.Utils.Socket.MulticastSenderSocket;
import com.CxWave.Wave.Resources.ResourceEnums.ResourceId;
import com.CxWave.Wave.Resources.ResourceEnums.TraceLevel;

@OwnerOM (om = LightHouseTransportObjectManager.class)
@Cardinality (1)
@WorkerPriority (WaveWorkerPriority.WAVE_WORKER_PRIORITY_0)
public class LightHouseTransportBroadcastLightPulseWorker extends WaveWorker
{
    private final MulticastSenderSocket m_multicastSenderSocket;

    public LightHouseTransportBroadcastLightPulseWorker (final LightHouseTransportObjectManager lightHouseTransportObjectManager)
    {
        super (lightHouseTransportObjectManager);

        m_multicastSenderSocket = new MulticastSenderSocket (LightHouseToolKit.getLightHouseMulticastIpAddress (), LightHouseToolKit.getLightHouseMulticastPort ());

        waveAssert (null != m_multicastSenderSocket);
        waveAssert (m_multicastSenderSocket.isValid ());
    }

    private void broadcastLightPulseMessageHandler (final LightHouseTransportBroadcastLightPulseMessage lightHouseTransportBroadcastLightPulseMessage)
    {
        trace (TraceLevel.TRACE_LEVEL_DEVEL, "LightHouseTransportBroadcastLightPulseWorker.broadcastLightPulseMessageHandler : Entering ...");

        waveAssert (null != lightHouseTransportBroadcastLightPulseMessage);

        final LightPulse lightPulse = lightHouseTransportBroadcastLightPulseMessage.getLightPulse ();

        waveAssert (null != lightPulse);

        final StringBuffer serializedLightPulseStringBuffer = new StringBuffer ();

        lightPulse.serializeTo (serializedLightPulseStringBuffer);

        m_multicastSenderSocket.send (serializedLightPulseStringBuffer);

        lightHouseTransportBroadcastLightPulseMessage.setCompletionStatus (ResourceId.WAVE_MESSAGE_SUCCESS);
        reply (lightHouseTransportBroadcastLightPulseMessage);
    }
}
